// @cloudinary/transformation-builder-sdk/qualifiers/background/shared/BackgroundGenerativeFillQualifier.js
import { BackgroundQualifier } from "./base/BackgroundQualifier.js";
/**
 * @description Automatically fills the padded area using generative AI to extend the image seamlessly.
 * Optionally include a prompt to guide the image generation.
 * @memberOf Qualifiers.Background
 */
class BackgroundGenerativeFillQualifier extends BackgroundQualifier {
    constructor(backgroundValue) {
        super(backgroundValue);
        this._prompt = getPromptFromBackgroundValue(backgroundValue);
        this._backgroundType = 'generativeFill';
    }
    getPrompt() {
        return this._prompt;
    }
    getBackgroundType() {
        return this._backgroundType;
    }
    /**
     * @description
     * Override the toString() function to explicitly stringify the qualifier.
     */
    toString() {
        return `b_gen_fill${this._prompt ? `:prompt_${this._prompt}` : ''}`;
    }
    static createInstanceFromModel(backgroundModel) {
        const backgroundValue = `gen_fill${backgroundModel.prompt ? `:prompt_${backgroundModel.prompt}` : ''}`;
        return new BackgroundGenerativeFillQualifier(backgroundValue);
    }
}
/**
 * @description Helper for checking if the background value is type of Generative Fill.
 */
function isGenerativeFillBackgroundValue(backgroundValue) {
    return typeof backgroundValue === 'string' && backgroundValue.startsWith('gen_fill');
}
/**
 * @description Helper for getting prompt from the Generative Fill background value.
 */
function getPromptFromBackgroundValue(backgroundValue) {
    return backgroundValue.split(':prompt_')[1];
}
export { BackgroundGenerativeFillQualifier, isGenerativeFillBackgroundValue, getPromptFromBackgroundValue };
